File: internal/api/handlers/telemetry.go
package handlers

import (
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"go.uber.org/zap"

	"github.com/getlawrence/lawrence-oss/internal/storage"
)

// TelemetryHandlers handles telemetry-related API endpoints
type TelemetryHandlers struct {
	storage *storage.Container
	logger  *zap.Logger
}

// NewTelemetryHandlers creates a new telemetry handlers instance
func NewTelemetryHandlers(storage *storage.Container, logger *zap.Logger) *TelemetryHandlers {
	return &TelemetryHandlers{
		storage: storage,
		logger:  logger,
	}
}

// QueryMetricsRequest represents the request for querying metrics
type QueryMetricsRequest struct {
	AgentID     *string    `json:"agent_id" binding:"omitempty,uuid"`
	GroupID     *string    `json:"group_id" binding:"omitempty,uuid"`
	MetricName  *string    `json:"metric_name"`
	StartTime   time.Time  `json:"start_time" binding:"required"`
	EndTime     time.Time  `json:"end_time" binding:"required"`
	Limit       int        `json:"limit"`
	UseRollups  bool       `json:"use_rollups"`
}

// QueryLogsRequest represents the request for querying logs
type QueryLogsRequest struct {
	AgentID     *string    `json:"agent_id" binding:"omitempty,uuid"`
	GroupID     *string    `json:"group_id" binding:"omitempty,uuid"`
	Severity    *string    `json:"severity"`
	Search      *string    `json:"search"`
	StartTime   time.Time  `json:"start_time" binding:"required"`
	EndTime     time.Time  `json:"end_time" binding:"required"`
	Limit       int        `json:"limit"`
}

// QueryTracesRequest represents the request for querying traces
type QueryTracesRequest struct {
	AgentID     *string    `json:"agent_id" binding:"omitempty,uuid"`
	GroupID     *string    `json:"group_id" binding:"omitempty,uuid"`
	TraceID     *string    `json:"trace_id"`
	ServiceName *string    `json:"service_name"`
	StartTime   time.Time  `json:"start_time" binding:"required"`
	EndTime     time.Time  `json:"end_time" binding:"required"`
	Limit       int        `json:"limit"`
}

// TelemetryOverviewResponse represents the telemetry overview
type TelemetryOverviewResponse struct {
	TotalMetrics    int64     `json:"totalMetrics"`
	TotalLogs       int64     `json:"totalLogs"`
	TotalTraces     int64     `json:"totalTraces"`
	ActiveAgents    int       `json:"activeAgents"`
	Services        []string  `json:"services"`
	LastUpdated     time.Time `json:"lastUpdated"`
}

// ServicesResponse represents the services list
type ServicesResponse struct {
	Services []string `json:"services"`
	Count    int      `json:"count"`
}

// handleQueryMetrics handles POST /api/v1/telemetry/metrics/query
func (h *TelemetryHandlers) HandleQueryMetrics(c *gin.Context) {
	// Not implemented in current interface
	c.JSON(http.StatusNotImplemented, gin.H{"error": "Metrics query not implemented"})
}

// handleQueryLogs handles POST /api/v1/telemetry/logs/query
func (h *TelemetryHandlers) HandleQueryLogs(c *gin.Context) {
	// Not implemented in current interface
	c.JSON(http.StatusNotImplemented, gin.H{"error": "Logs query not implemented"})
}

// handleQueryTraces handles POST /api/v1/telemetry/traces/query
func (h *TelemetryHandlers) HandleQueryTraces(c *gin.Context) {
	// Not implemented in current interface
	c.JSON(http.StatusNotImplemented, gin.H{"error": "Traces query not implemented"})
}

// handleGetTelemetryOverview handles GET /api/v1/telemetry/overview
func (h *TelemetryHandlers) HandleGetTelemetryOverview(c *gin.Context) {
	// Not implemented in current interface
	c.JSON(http.StatusNotImplemented, gin.H{"error": "Telemetry overview not implemented"})
}

// handleGetServices handles GET /api/v1/telemetry/services
func (h *TelemetryHandlers) HandleGetServices(c *gin.Context) {
	// Not implemented in current interface
	c.JSON(http.StatusNotImplemented, gin.H{"error": "Services list not implemented"})
}

// parseTimeRange parses time range string to duration
func parseTimeRange(timeRange string) (time.Duration, error) {
	switch timeRange {
	case "5m":
		return 5 * time.Minute, nil
	case "15m":
		return 15 * time.Minute, nil
	case "1h":
		return 1 * time.Hour, nil
	case "6h":
		return 6 * time.Hour, nil
	case "24h":
		return 24 * time.Hour, nil
	case "7d":
		return 7 * 24 * time.Hour, nil
	default:
		// Try to parse as duration string
		return time.ParseDuration(timeRange)
	}
}

File: ui/src/api/telemetry.ts
import { apiGet, apiPost } from "./base";

export interface TelemetryOverview {
  totalMetrics: number;
  totalLogs: number;
  totalTraces: number;
  activeAgents: number;
  services: string[];
  lastUpdated: string;
}

export interface ServicesResponse {
  services: string[];
  count: number;
}

export interface MetricQueryRequest {
  agent_id?: string;
  group_id?: string;
  metric_name?: string;
  start_time: string;
  end_time: string;
  limit?: number;
  use_rollups?: boolean;
}

export interface LogQueryRequest {
  agent_id?: string;
  group_id?: string;
  severity?: string;
  search?: string;
  start_time: string;
  end_time: string;
  limit?: number;
}

export interface TraceQueryRequest {
  agent_id?: string;
  group_id?: string;
  trace_id?: string;
  service_name?: string;
  start_time: string;
  end_time: string;
  limit?: number;
}

export interface MetricData {
  timestamp: string;
  agent_id: string;
  group_id?: string;
  service_name: string;
  metric_name: string;
  value: number;
  metric_attributes: Record<string, any>;
}

export interface LogData {
  timestamp: string;
  agent_id: string;
  group_id?: string;
  service_name: string;
  severity_text: string;
  severity_number: number;
  body: string;
  trace_id?: string;
  span_id?: string;
  log_attributes: Record<string, any>;
}

export interface TraceData {
  timestamp: string;
  agent_id: string;
  group_id?: string;
  trace_id: string;
  span_id: string;
  parent_span_id?: string;
  service_name: string;
  span_name: string;
  duration: number;
  status_code: string;
  span_attributes: Record<string, any>;
}

export interface MetricsQueryResponse {
  metrics: MetricData[];
  count: number;
}

export interface LogsQueryResponse {
  logs: LogData[];
  count: number;
}

export interface TracesQueryResponse {
  traces: TraceData[];
  count: number;
}

// Get telemetry overview
export const getTelemetryOverview = (): Promise<TelemetryOverview> => {
  return apiGet<TelemetryOverview>("/telemetry/overview");
};

// Get services list
export const getServices = (): Promise<ServicesResponse> => {
  return apiGet<ServicesResponse>("/telemetry/services");
};

// Query metrics
export const queryMetrics = (
  request: MetricQueryRequest,
): Promise<MetricsQueryResponse> => {
  return apiPost<MetricsQueryResponse>("/telemetry/metrics/query", request);
};

// Query logs
export const queryLogs = (
  request: LogQueryRequest,
): Promise<LogsQueryResponse> => {
  return apiPost<LogsQueryResponse>("/telemetry/logs/query", request);
};

// Query traces
export const queryTraces = (
  request: TraceQueryRequest,
): Promise<TracesQueryResponse> => {
  return apiPost<TracesQueryResponse>("/telemetry/traces/query", request);
};
